tardis/util/base.py

```diff
@@ -699,6 +699,14 @@ def update_packet_pbar(i, current_iteration, no_of_packets, total_iterations):
     packet_pbar.update(i)
 
 
+def refresh_packet_pbar():
+    """
+    Refresh packet progress bar after each iteration.
+
+    """
+    packet_pbar.refresh()
+
+
```

Comment:
Moving this inside the above function may do the trick. Please let me know if that works. Also add a single line comment for the same instead of having a function.

Reply:
Since the update/refresh does not occur only at the end of each iteration, I decided to call it when the iteration ends. Adding the `packet_pbar.refresh()` at every packet would increase the simulation time severely (around 1 min/iteration) in contrast to refreshing only at the end (70s for 20 iterations on my system).

This is why I could not find a suitable place to add the refresh anywhere in the function above  (in `update_packet_pbar()`). Also because one cannot conclude when the final packet is sent (probably because it is numba jitted in parallel mode).

Thus, the only remaining solution was to either add the `packet_bar.refresh()` within the `update_iterations_pbar()` function as it is called at the end of every iteration, or have a function of its own.

Reply:
> Thus, the only remaining solution was to either add the `packet_bar.refresh()` within the `update_iterations_pbar()` function as it is called at the end of every iteration, or have a function of its own.

I suppose it is better to have just this line `packet_bar.refresh()` after the `update_iterations_pbar(1)` as having a function just for this is not good, in my opinion, and is just overhead with a call within a functional call.

Reply:
> I suppose it is better to have just this line `packet_bar.refresh()` after the `update_iterations_pbar(1)` as having a function just for this is not good, in my opinion, and is just overhead with a call within a functional call.

The packet_pbar and iterations_pbar instances of the progress bars are initialized and processed in `tardis/utility/base.py`. So, `packet_bar.refresh()` after `update_iterations_pbar(1)` would require instantiating the progress bars in `montecarlo_numba/base.py` as well, which may not be the best approach.

I agree that it is an avoidable overhead, in which case I can just add it in the `update_iterations_pbar()` method after `iterations_pbar.update(i)`. It would look something like:

```
def update_iterations_pbar(i):
    iterations_pbar.update(i)
    packet_bar.refresh()
```

Reply:
This would be inefficient as the current implementation updates the progress bar by 1 per cent after a specific increase in the progress as the packets are propagated. What this may do is that after every 1 per cent, the progress bar will get refreshed and updated. For this, I suppose you can look into making something like this: (pseudo code here)
```
If the progress is above let's say, 95% (would be good if this is as close to 100 as possible or maybe even 100):
  refresh the bar for each iteration until it reaches the end,
else:
  keep doing what it initially was doing,
```
What happens with the `numba jitted` functions is that the information is present with the compiler (machine level compiler) but is not actually passed onto into Python runtime (Python interpreter), which causes loss of information as hence it doesn't understand how to add all the multithreaded information together from different cores. 

Another possible solution could be that you update the progress bar by returning the instance of the progress bar from the `update_iterations_pbar()` function and then update it when all the progress is finished. What I mean by this is that even when we are running TARDIS in multithreaded mode, the progress bar takes into consideration all the progress made on all the cores. And in the end, please correct me, it is still compiled on a single core. So a single call for refreshing the progress bar at the end after the completion of all the propagated packets may do the trick. It works for all other iterations, expect the last one. This would be more efficient.

Reply:
> This would be inefficient as the current implementation updates the progress bar by 1 per cent after a specific increase in the progress as the packets are propagated. What this may do is that after every 1 per cent, the progress bar will get refreshed and updated. 

The `update_iterations_pbar()` is called to update the _iterations_ progress bar (`iterations_pbar`) at the end of each iteration, only after the _packet_ progress bar (`packet_pbar`) has finished updating for that iteration, i.e., after all the packets for that iteration have been sent. So I think this refresh would not interfere with the `packet_pbar` while the packets are being sent and only update it after sending 100% of the packets.

> What happens with the numba jitted functions is that the information is present with the compiler (machine level compiler) but is not actually passed onto into Python runtime (Python interpreter), which causes loss of information as hence it doesn't understand how to add all the multithreaded information together from different cores.

Thank you for the clear description. This may be why there is no possible way to fetch the completion percentage as the number of packets sent through different cores is not trackable. 

> So a single call for refreshing the progress bar at the end after the completion of all the propagated packets may do the trick.

My intention with the code block mentioned in the previous reply was precisely this. Since  `update_iterations_pbar()` is called at the end of every iteration, adding the refresh part within it would also be subject to a call at the end of each iteration when all the packets have propagated. Please correct me if I've misunderstood.